docs/proposals/multipart_trasnfer_mode.md

```diff
@@ -0,0 +1,296 @@
+* Do not require any state to be maintained in the server side
+
+## High Level Protocol Specs
+* The name of the transfer is `multipart`
+* Batch requests are the same as `basic` requests except that `{"transfers": ["multipart", "basic"]}` is the
+  expected transfers value. Clients MUST retain `basic` as the fallback transfer mode to ensure compatiblity with
+  servers not implementing this extension.
+* `{"operation": "download"}` replies work exactly like `basic` download request with no change
+* `{"operation": "upload"}` replies will break the upload into several `actions`:
+  * `parts` (optional), a list of zero or more part upload actions
+  * `verify` (optional), an action to verify the file is in storage, similar to `basic` upload verify actions
+  * `abort` (optional), an action to abort the upload and clean up all unfinished chunks and state
+* Just like `basic` transfers, if the file fully exists and is committed to storage, no `actions` will be provided
+  in the reply and the upload can simply be skipped
+* If a `verify` action is provided, calling it is required and not optional. In some cases, this endpoint may be used
+  to finalize the upload.
+* An empty or missing list of `parts` with a `verify` action may mean all parts have been uploaded but `verify` still
+  needs to be called by the client.
+* Authentication and authorization behave just like with the `basic` protocol.
+
+## Action Objects
+Each one of the `parts` and the `verify` and `abort` actions contain instructions for the client on how to send a
+request performing a the particular action. These are similar to `basic` transfer adapter `actions` but may include
+some common, as well as action specific additional parameters.
+
+All actions allow `href`, `header` and `expires_in` parameters just like `basic` transfer actions.
+
+### `parts` actions
+Each `parts` action should include the `pos` and `size` attributes, in addition to the attributes specified
+above:
+* `pos` indicate the position in bytes within the file in which the part should begin. If not specified, `0` (that is the
+beginning of the file) is assumed.
+* `size` is the size of the part in bytes. If `size` is omitted, default to read until the end of file.
+* If both `pos` and `size` are omitted, the action is expected to be a single-part upload of the entire file
+
+In adition, `parts` actions may include the following parameters:
+* `method`, with `PUT` as the default method if none is specified. This allows customizing the HTTP method used when uploading
+object parts.
+* `want_digest` to specify an expected HTTP `Digest` header, as described below.
+
+### `verify` action
+The `verify` action is similar to `basic` transfer mode `verify`, with the following additional parameters:
+* `params` - an object with additional parameters to send to the server as part of the `POST` JSON body.
+  These parameters are to be sent as provided back to the server.
```

Comment:
Can we specify that this `params` structure is passed as a `params` structure?  That is, that our response looks like this:

```
{
  "params": {
    "a": 256,
    "b": "string",
    "c": null
  }
}
```

Doing so means we can add data in the future easier and it also means that implementing parsing of this field in strongly typed languages like Go and Rust is a _lot_ easier than if we insert them randomly into the top-level body.